Authentication with jwt

utsabk opened the pull request:
Authentication with Jsonwebtoken
     - Add signup and signin
     - sign a jwt for a signed in user
     - save token in cookies
     - verify user with token from cookies

All the request queries are listed in request.rest which allows to make post requests directly form the code editor 

close #5

utsabk committed:
Initialize Node Express app

utsabk committed:
Add environment variables

utsabk committed:
Authentication with Jsonwebtoken
 - Add signup and signin
 - sign a jwt for a signed in user
 - save token in cookies
 - verify user with token from cookies

utsabk committed:
Remove cookies
 jwt is not saved in cookies rather sent in every request

utsabk commented:
Cookies removed
- jwt is sent back in response body when signed in
- authorized requests are verified with signed jwt

awetg commented on index.js:
```diff
@@ -0,0 +1,94 @@
+    const token = jwt.sign(currentUser, process.env.SECRET_KEY, {
+      expiresIn: 60 * 15,
+    });
+    delete currentUser.password;
```
The deletion of password must happen before signing the JWT, the idea of never sending user password to anyone is very basic protection, so have to be removed from the JWT payload.

awetg commented on index.js:
```diff
@@ -0,0 +1,94 @@
+      message: 'User Already Exists! Login or choose another user email',
+    });
+  } else {
+    const userId = generateId(Users);
```
`generateId()` function is over glorified version of `Users.length + 1`. I can't see why it needs to be like that, it would be nice if you can clarify this.

utsabk replied:
correct, for this scenario where **user** is stored in a variable `Users.length + 1` works perfectly fine. The idea behind `generateId()` is to make a universal function that would work in any condition. for instance when a user is deleted from **Users** list, in such case `Users.length` might not give us the right number

awetg replied:
`Users.length` should always give us the correct number of users, the fact that we added users or deleted users should not matter. The value `Users.length` should always match the current number of users.

utsabk replied:
`Users.length` gives us the no. of **users** in the list. However, if we delete a **user** from somewhere in the middle, assuming the **ids** are not updating on its own, on adding a new **user** to the list again, it will duplicate the **user id**.

awetg replied:
Okay, I see the problem now then we will have to move to using GUID as user id, we can use a simple GUID generator function for the time being instead of using `uuid` package.

bishwasshrestha replied:
i think the problem with using users.length+1 method is that if u delete something from the middle it makes two posts with same id. {
        "id": 6,
        "title": "title 7",
        "note": "i m added note",
        "dateCreated": "today"
    },
    {
        "id": 6,
        "title": "title 8",
        "note": "i m added note",
        "dateCreated": "today"
    }

Code: 
if(!body.id){
    console.log('no id on this post');
    const newId = Notes.length + 1;
    const newNote = {
    id: newId,
    title: body.title,
    note: body.note,
    dateCreated: body.dateCreated
    };

    Notes.push(newNote);
  }
  else
    Notes.push(body);

awetg replied:
Yes, this problem also applies to notes. Let's use UUID for both users and notes.

awetg reviewed (commented):
In general it look good to me.

awetg commented on requests.rest:
```diff
@@ -0,0 +1,22 @@
+GET / HTTP/1.1
```
Can you remove this file? `requests.rest` it looks like a test file using some VSCODE extension, since it is not relevant to the project and we will introduce other ways to test the API, maybe we can add `.rest` files to `.gitignore`

utsabk replied:
it's a file for **REST Client**, a **vscode** extension. It contains all the tets requests, and it's very handy for testing POST requests which are not possible from the browser, for that we need to switch to another application like **Postman**. Since it is only used in the development, nothing to do with API, I will add to `.gitignore`

utsabk committed:
Replace ID with UUID
 - general id is replaced with uuid
 - password is deleted before signing the JWT
